File: WdtcCore/src/main/java/org/wdt/wdtc/download/quilt/QuiltDownloadInfo.java
package org.wdt.wdtc.download.quilt;

import lombok.Getter;
import org.wdt.wdtc.download.infterface.DownloadInfo;
import org.wdt.wdtc.download.infterface.InstallTask;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.manger.FileManger;
import org.wdt.wdtc.utils.gson.JSONObject;
import org.wdt.wdtc.utils.gson.JSONUtils;

import java.io.File;
import java.io.IOException;

public class QuiltDownloadInfo implements DownloadInfo {
    private static final String LibraryListUrl = "https://meta.quiltmc.org/v3/versions/loader/%s/%s/profile/json";

    protected final Launcher launcher;
    @Getter
    protected final String QuiltVersionNumber;

    public QuiltDownloadInfo(Launcher launcher, String quiltVersionNumber) {
        this.launcher = launcher;
        QuiltVersionNumber = quiltVersionNumber;
    }

    public File getQuiltVersionJson() {
        return new File(FileManger.getWdtcCache() + "/" + launcher.getVersionNumber() + "-quilt-" + QuiltVersionNumber + ".json");
    }

    public String getQuiltVersionJsonUrl() {
        return String.format(LibraryListUrl, launcher.getVersionNumber(), QuiltVersionNumber);
    }

    public JSONObject getQuiltGameVersionJsonObject() throws IOException {
        return JSONUtils.getJSONObject(getQuiltVersionJson());
    }

    @Override
    public String getModVersion() {
        return QuiltVersionNumber;
    }

    @Override
    public InstallTask getModInstallTask() {
        return new QuiltInstallTask(launcher, QuiltVersionNumber);
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/download/quilt/QuiltDownloadInfo.kt
package org.wdt.wdtc.core.impl.download.quilt

import com.google.gson.annotations.JsonAdapter
import org.wdt.utils.gson.readFileToClass
import org.wdt.wdtc.core.openapi.download.interfaces.ModDownloadInfoInterface
import org.wdt.wdtc.core.openapi.download.interfaces.ModInstallTaskInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.game.serializeGameRuntimeDependencyGsonBuilder
import org.wdt.wdtc.core.openapi.manger.KindOfMod
import org.wdt.wdtc.core.openapi.manger.officialDownloadSource
import org.wdt.wdtc.core.openapi.manger.wdtcCache
import org.wdt.wdtc.core.openapi.utils.gson.DownloadInfoTypeAdapter
import java.io.File

@JsonAdapter(DownloadInfoTypeAdapter::class)
open class QuiltDownloadInfo(
	protected val version: Version,
	override val modVersion: String
) : ModDownloadInfoInterface {
  constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) :
	  this(version, versionsJsonObjectInterface.versionNumber)
	
	
	private val libraryListUrl = "${officialDownloadSource.quiltMetaUrl}v3/versions/loader/%s/%s/profile/json"
	
	val quiltVersionJson: File
    get() = File(wdtcCache, "${version.versionNumber}-quilt-$modVersion.json")
	
	val quiltVersionJsonUrl: String
    get() = libraryListUrl.format(version.versionNumber, modVersion)
	
	val quiltProfileJsonObject: QuiltProfileJsonObject
		get() = quiltVersionJson.readFileToClass(serializeGameRuntimeDependencyGsonBuilder)
  override val modInstallTask: ModInstallTaskInterface
    get() = QuiltInstallTask(version, modVersion)
  override val modKind: KindOfMod
    get() = KindOfMod.QUILT

  override fun toString(): String {
    return "QuiltDownloadInfo(modVersion='$modVersion')"
  }
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is QuiltDownloadInfo) return false
		
		if (modVersion != other.modVersion) return false
		
		return true
	}
	
	override fun hashCode(): Int {
		return modVersion.hashCode()
	}
	
	
}
